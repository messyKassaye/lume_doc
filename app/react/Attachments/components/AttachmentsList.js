import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { advancedSort } from 'app/utils/advancedSort';
import { Translate } from 'app/I18N';
import { NeedAuthorization } from 'app/Auth';
import Attachment from 'app/Attachments/components/Attachment';

import UploadSupportingFile from './UploadSupportingFile';

export default class AttachmentsList extends Component {
  static arrangeFiles(files = []) {
    return advancedSort(files, { property: 'originalname' });
  }

  render() {
    const { parentId, parentSharedId, readOnly, storeKey } = this.props;
    const forcedReadOnly = readOnly || Boolean(this.props.isTargetDoc);

    let uploadAttachmentButton = null;
    if (!this.props.isTargetDoc) {
      uploadAttachmentButton = (
        <NeedAuthorization roles={['admin', 'editor']}>
          <div className="attachment-add">
            <UploadSupportingFile entitySharedId={this.props.parentSharedId} storeKey={storeKey} />
          </div>
        </NeedAuthorization>
      );
    }

    const attachments = AttachmentsList.arrangeFiles(this.props.attachments);
    return (
      <div className="attachments-list-parent">
        <div className="attachments-list-header">
          <h2>
            <Translate>Supporting files</Translate>
          </h2>
          {uploadAttachmentButton}
        </div>
        <div className="attachments-list">
          {attachments.map((file, index) => (
            <Attachment
              key={index}
              file={file}
              parentId={parentId}
              readOnly={forcedReadOnly}
              storeKey={storeKey}
              parentSharedId={parentSharedId}
              isSourceDocument={false}
            />
          ))}
        </div>
      </div>
    );
  }
}

AttachmentsList.defaultProps = {
  attachments: [],
  readOnly: false,
  isTargetDoc: false,
  parentId: null,
  parentSharedId: null,
  storeKey: '',
};

AttachmentsList.propTypes = {
  attachments: PropTypes.arrayOf(PropTypes.object),
  parentId: PropTypes.string,
  parentSharedId: PropTypes.string,
  readOnly: PropTypes.bool,
  isTargetDoc: PropTypes.bool,
  storeKey: PropTypes.string,
};

AttachmentsList.contextTypes = {
  confirm: PropTypes.func,
};
