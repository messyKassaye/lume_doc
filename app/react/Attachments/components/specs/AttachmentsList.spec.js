import React from 'react';
import { shallow } from 'enzyme';
import { fromJS as Immutable } from 'immutable';

import AttachmentsList from '../AttachmentsList';

describe('AttachmentsList', () => {
  let component;
  let props;
  let files;

  beforeEach(() => {
    files = Immutable([
      { originalname: 'Human name 1', filename: 'filename1.ext' },
      { originalname: 'A Human name 2', filename: 'filename2.ext' },
    ]);

    props = {
      files,
      parentId: 'parentId',
      parentSharedId: 'parentSharedId',
      isDocumentAttachments: false,
      readOnly: false,
      storeKey: 'storeKey',
    };
  });

  const render = () => {
    component = shallow(<AttachmentsList {...props} />);
  };

  it('should render a sorted list of attachments (files)', () => {
    render();
    expect(component).toMatchSnapshot();
  });

  describe('When parent is Target Document', () => {
    beforeEach(() => {
      props.isTargetDoc = true;
      render();
    });

    it('should treat all Attachments as read only', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('when files is empty', () => {
    it('should render nothing if user not logged in', () => {
      props.files = Immutable([]);
      render();
      expect(component).toMatchSnapshot();
    });

    it('should add button in Downloads section', () => {
      props.files = Immutable([]);
      props.user = Immutable({ _id: 'user' });
      render();
      expect(component).toMatchSnapshot();
    });
  });
});
