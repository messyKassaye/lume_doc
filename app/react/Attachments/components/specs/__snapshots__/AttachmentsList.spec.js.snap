// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentsList When parent is Target Document should treat all Attachments as read only 1`] = `
<div
  className="attachments-list-parent"
>
  <div
    className="attachments-list-header"
  >
    <h2>
      <Connect(Translate)>
        Supporting files
      </Connect(Translate)>
    </h2>
  </div>
  <div
    className="attachments-list"
  />
</div>
`;

exports[`AttachmentsList should render a sorted list of attachments (files) 1`] = `
<div
  className="attachments-list-parent"
>
  <div
    className="attachments-list-header"
  >
    <h2>
      <Connect(Translate)>
        Supporting files
      </Connect(Translate)>
    </h2>
    <Connect(NeedAuthorization)
      roles={
        Array [
          "admin",
          "editor",
        ]
      }
    >
      <div
        className="attachment-add"
      >
        <Connect(UploadSupportingFile)
          entitySharedId="parentSharedId"
          storeKey="storeKey"
        />
      </div>
    </Connect(NeedAuthorization)>
  </div>
  <div
    className="attachments-list"
  />
</div>
`;

exports[`AttachmentsList when files is empty should add button in Downloads section 1`] = `
<div
  className="attachments-list-parent"
>
  <div
    className="attachments-list-header"
  >
    <h2>
      <Connect(Translate)>
        Supporting files
      </Connect(Translate)>
    </h2>
    <Connect(NeedAuthorization)
      roles={
        Array [
          "admin",
          "editor",
        ]
      }
    >
      <div
        className="attachment-add"
      >
        <Connect(UploadSupportingFile)
          entitySharedId="parentSharedId"
          storeKey="storeKey"
        />
      </div>
    </Connect(NeedAuthorization)>
  </div>
  <div
    className="attachments-list"
  />
</div>
`;

exports[`AttachmentsList when files is empty should render nothing if user not logged in 1`] = `
<div
  className="attachments-list-parent"
>
  <div
    className="attachments-list-header"
  >
    <h2>
      <Connect(Translate)>
        Supporting files
      </Connect(Translate)>
    </h2>
    <Connect(NeedAuthorization)
      roles={
        Array [
          "admin",
          "editor",
        ]
      }
    >
      <div
        className="attachment-add"
      >
        <Connect(UploadSupportingFile)
          entitySharedId="parentSharedId"
          storeKey="storeKey"
        />
      </div>
    </Connect(NeedAuthorization)>
  </div>
  <div
    className="attachments-list"
  />
</div>
`;
